python/ray/serve/controller.py

```diff
@@ -129,8 +132,60 @@ def get_http_proxies(self) -> Dict[NodeId, ActorHandle]:
         """Returns a dictionary of node ID to http_proxy actor handles."""
         return self.http_state.get_http_proxy_handles()
 
+    async def update_autoscaling(self) -> None:
+        for deployment_name, (backend_info,
+                              route_prefix) in self.list_deployments().items():
+            backend_config = backend_info.backend_config
+            autoscaling_config = backend_config.autoscaling_config
+            if autoscaling_config is None:
+                continue
+            replicas = self.backend_state_manager._backend_states[
+                deployment_name]._replicas
+            running_replicas = replicas.get([ReplicaState.RUNNING])
+            current_num_ongoing_requests = []
+            for replica in running_replicas:
+                replica_tag = replica.replica_tag
+                num_ongoing_requests = (
+                    self.autoscaling_metrics_store.window_average(
+                        replica_tag,
+                        time.time() - autoscaling_config.look_back_period_s))
+                current_num_ongoing_requests.append(num_ongoing_requests)
+            if len(current_num_ongoing_requests) == 0:
+                print("length of list of num requests is zero")
+                continue
+            print("current ongoing num requests: ",
+                  current_num_ongoing_requests)
+            new_backend_config = backend_config.copy()
+            new_backend_config.num_replicas = calculate_desired_num_replicas(
+                autoscaling_config, current_num_ongoing_requests)
+            print("desired num replicas: ", new_backend_config.num_replicas)
+            print("Metrics dump:  ",
+                  self._dump_autoscaling_metrics_for_testing())
+            # continue #TODO: remove
+            replica_config = backend_info.replica_config
+            deployer_job_id = backend_info.deployer_job_id
+            backend_config_proto_bytes = new_backend_config.to_proto_bytes()
+            # TODO: Only allow autoscaling for versioned deployments for now
+            # TODO: Weird behavior: when spamming a new deployment, the
+            # num ongoing requests goes from 0, to 0.5, to 0.66, to 1, to
+            # (huge) (15-20ish)
+            goal_id, updating = await self.deploy(
+                deployment_name,
+                backend_config_proto_bytes,
+                replica_config,
+                version=backend_info.version,
+                prev_version=backend_info.version,
+                route_prefix=route_prefix,
+                deployer_job_id=deployer_job_id)
+
     async def run_control_loop(self) -> None:
         while True:
+            # try:
+            await self.update_autoscaling()
```

Comment:
I wonder whether or not we should start move the control loop into asyncio task based instead of processing each sequentially. Here is what I had in mind:
```
async def run_control_loop(self):
    while True:
        tasks = [asyncio.create_task(coro) for coro in [self.update_autoscaling(), self.http_state.update(), self.backend_state.update()]]
        await asyncio.sleep()
        # in next iteration check whether tasks are completed, if not, skip creating new ones.
```
@jiaodong  @edoakes thoughts?

Reply:
parallelizing makes sense for optimization but i think we should also be aware of and be explicit about order of execution ? For example if it's more desirable to only update http_state after backend_state changed in case it's deploying or changing new deployment ? Similarly what will happen if these update() have different interval with interleaved cycles.

I'm not clear what is expected to happen in update_autoscaling() yet but i saw some deploy_backend calls there